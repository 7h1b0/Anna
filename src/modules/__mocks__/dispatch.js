const dispatch = jest.fn(async () => {});
export default dispatch;
