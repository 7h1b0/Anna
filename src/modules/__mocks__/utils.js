export const isBankHoliday = jest.fn(() => false);
