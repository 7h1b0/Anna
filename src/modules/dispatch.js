import { HUE_LIGHT, DIO, SCENE } from './type';
import { findById as findSceneById } from './models/scene';
import { setLightState } from '../services/hueService';
import dioAdd from '../services/dioService';

export default function dispatch({ type, id, body }) {
  switch (type) {
    case HUE_LIGHT:
      return setLightState(id, body);

    case DIO:
      return dioAdd(id, body.on);

    case SCENE:
      return findSceneById(id).then(scene => {
        if (scene) {
          return Promise.all(
            scene.actions.map(({ type, targetId, body }) =>
              dispatch({ type, id: targetId, body }),
            ),
          );
        }
        throw new Error(`Cannot find scene ${id}`);
      });

    default:
      return Promise.reject(`Unknow type ${type}`);
  }
}
