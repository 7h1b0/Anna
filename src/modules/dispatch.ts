import TYPES from './type';
import { AnnaAction } from './actions';
import { findById as findSceneById } from './models/scene';
import { setLightState } from 'services/hueService';
import dioAdd from 'services/dioService';

export default async function dispatch(action: AnnaAction): Promise<void> {
  switch (action.type) {
    case TYPES.HUE_LIGHT:
      return setLightState(action.targetId, action.body);

    case TYPES.DIO:
      return dioAdd(action.targetId, action.body.on);

    case TYPES.SCENE:
      await findSceneById(action.targetId).then(scene => {
        if (scene) {
          return Promise.all(scene.actions.map(action => dispatch(action)));
        }
        throw new Error(`Cannot find scene ${action.targetId}`);
      });
  }
}
