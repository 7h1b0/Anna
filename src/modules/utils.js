export function omit(object, props = []) {
  return Object.keys(object).reduce((acc, key) => {
    if (!props.includes(key)) {
      acc[key] = object[key];
    }
    return acc;
  }, {});
}

/**
 * Source
 * http://techneilogy.blogspot.fr/2012/02/couple-of-years-ago-i-posted-source.html
 */
export function isBankHoliday(timestamp) {
  function getPublicHolidays(year) {
    const a = year % 19;
    const b = Math.floor(year / 100);
    const c = year % 100;
    const d = Math.floor(b / 4);
    const e = b % 4;
    const f = Math.floor((b + 8) / 25);
    const g = Math.floor((b - f + 1) / 3);
    const h = (19 * a + b - d - g + 15) % 30;
    const i = Math.floor(c / 4);
    const k = c % 4;
    const l = (32 + 2 * e + 2 * i - h - k) % 7;
    const m = Math.floor((a + 11 * h + 22 * l) / 451);
    const n0 = h + l + 7 * m + 114;
    const month = Math.floor(n0 / 31) - 1;
    const day = (n0 % 31) + 1;

    const jourDeLan = new Date(year, 0, 1);
    const lundiDePaques = new Date(year, month, day + 1);
    const feteDuTravail = new Date(year, 4, 1);
    const victoireDesAllies = new Date(year, 4, 8);
    const ascension = new Date(year, month, day + 39);
    const pentecote = new Date(year, month, day + 50);
    const feteNationale = new Date(year, 6, 14);
    const assomption = new Date(year, 7, 15);
    const toussaint = new Date(year, 10, 1);
    const armistice = new Date(year, 10, 11);
    const noel = new Date(year, 11, 25);

    return [
      jourDeLan,
      lundiDePaques,
      feteDuTravail,
      victoireDesAllies,
      ascension,
      pentecote,
      feteNationale,
      assomption,
      toussaint,
      armistice,
      noel,
    ];
  }

  const date = timestamp ? new Date(timestamp) : new Date();
  date.setHours(0, 0, 0, 0);
  return getPublicHolidays(date.getFullYear()).some(
    holiday => date.getTime() === holiday.getTime(),
  );
}
