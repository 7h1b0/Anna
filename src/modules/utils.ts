export function omit(object: object, props: string[] = []): object {
  return Object.keys(object).reduce((acc, key) => {
    if (!props.includes(key)) {
      acc[key] = object[key];
    }
    return acc;
  }, {});
}

/**
 * Source
 * http://techneilogy.blogspot.fr/2012/02/couple-of-years-ago-i-posted-source.html
 */
export function isBankHoliday(date: Date) {
  function getPublicHolidays(year: number): Date[] {
    const a = year % 19;
    const b = Math.floor(year / 100);
    const c = year % 100;
    const d = Math.floor(b / 4);
    const e = b % 4;
    const f = Math.floor((b + 8) / 25);
    const g = Math.floor((b - f + 1) / 3);
    const h = (19 * a + b - d - g + 15) % 30;
    const i = Math.floor(c / 4);
    const k = c % 4;
    const l = (32 + 2 * e + 2 * i - h - k) % 7;
    const m = Math.floor((a + 11 * h + 22 * l) / 451);
    const n0 = h + l + 7 * m + 114;
    const month = Math.floor(n0 / 31) - 1;
    const day = (n0 % 31) + 1;

    return [
      new Date(year, 0, 1),
      new Date(year, month, day + 1),
      new Date(year, 4, 1),
      new Date(year, 4, 8),
      new Date(year, month, day + 39),
      new Date(year, month, day + 50),
      new Date(year, 6, 14),
      new Date(year, 7, 15),
      new Date(year, 10, 1),
      new Date(year, 10, 11),
      new Date(year, 11, 25),
    ];
  }

  date.setHours(0, 0, 0, 0);
  return getPublicHolidays(date.getFullYear()).some(
    holiday => date.getTime() === holiday.getTime(),
  );
}
