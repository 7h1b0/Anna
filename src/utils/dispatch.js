const { HUE_GROUP, HUE_LIGHT, DIO, SCENE } = require('./type');
const Scene = require('../models/scene');
const { hueService, dioService } = require('../services/');

module.exports = function dispatch({ type, id, body }) {
  switch (type) {
    case HUE_GROUP:
      return hueService.setGroupState(id, body);

    case HUE_LIGHT:
      return hueService.setLightState(id, body);

    case DIO:
      return dioService.add(id, body.on);

    case SCENE:
      return Scene.findById(id).then((scene) => {
        if (scene) {
          return Promise.all(scene.actions.map(dispatch));
        }
        return Promise.reject(`Cannot find scene ${id}`);
      });

    default:
      return Promise.reject();
  }
};
