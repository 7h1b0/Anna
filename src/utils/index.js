const dispatch = require('./dispatch');
const cryptoUtil = require('./cryptoUtil');
const getJoiError = require('./errorUtil');
const tokenUtil = require('./tokenUtil');
const type = require('./type');

module.exports = {
  dispatch,
  cryptoUtil,
  getJoiError,
  tokenUtil,
  type,
};
