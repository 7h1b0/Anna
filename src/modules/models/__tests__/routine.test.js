import knex from '../../../knexClient';
import Routine from '../routine';
import dispatch from '../../dispatch';
import { isBankHoliday } from '../../utils';

jest
  .mock('../../dispatch')
  .mock('../../utils')
  .unmock('cron');

const initRoutines = [
  {
    routineId: '0fc1d78e-fd1c-4717-b610-65d2fa3d01b2',
    sceneId: '00c1d78e-fd1c-4717-b610-65d2fa3d01b2',
    name: 'test at 5am',
    interval: '0 5 * * *',
    enabled: true,
    runAtBankHoliday: true,
    createdBy: 'c10c80e8-49e4-4d6b-b966-4fc9fb98879f',
    createdAt: new Date('2018-01-01'),
    updatedAt: new Date('2018-01-02'),
  },
  {
    routineId: '1fc1d78e-fd1c-4717-b610-65d2fa3d01b2',
    sceneId: '10c1d78e-fd1c-4717-b610-65d2fa3d01b2',
    name: 'test at 9am',
    interval: '0 9 * * *',
    enabled: true,
    runAtBankHoliday: false,
    createdBy: 'c10c80e8-49e4-4d6b-b966-4fc9fb98879f',
    createdAt: new Date('2018-01-01'),
    updatedAt: new Date('2018-01-02'),
  },
];

describe('Routines', () => {
  beforeAll(async () => {
    await knex(Routine.TABLE).truncate();
  });

  beforeEach(async () => {
    await knex(Routine.TABLE).insert(initRoutines);
  });

  afterEach(async () => {
    await knex(Routine.TABLE).truncate();
  });

  describe('findAll', () => {
    it('should return all routines', async () => {
      const result = await Routine.findAll();
      expect(result).toMatchSnapshot();
    });
  });

  describe('findById', () => {
    it('should return only one routine', async () => {
      const result = await Routine.findById(initRoutines[0].routineId);
      expect(result).toMatchSnapshot();
    });

    it('should return undefined', async () => {
      const result = await Routine.findById(-1);
      expect(result).toBeUndefined();
    });
  });

  describe('save', () => {
    it('should save a new routine', async () => {
      const id = await Routine.save(
        'c10c80e8-49e4-4d6b-b966-4fc9fb98879f',
        'test_3',
        '20c1d78e-fd1c-4717-b610-65d2fa3d01b2',
        'schedule',
      );
      const routine = await knex(Routine.TABLE)
        .first()
        .where('routineId', id);

      expect(routine).toMatchSnapshot({
        routineId: expect.stringMatching(/[a-fA-F0-9-]{36}/),
        sceneId: expect.stringMatching(/[a-fA-F0-9-]{36}/),
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
      });
    });
  });

  describe('remove', () => {
    it('should remove a routine', async () => {
      const res = await Routine.remove(initRoutines[0].routineId);
      expect(res).toBe(1);

      const routine = await knex(Routine.TABLE)
        .first()
        .where('routineId', initRoutines[0].routineId);

      expect(routine).toBeUndefined();
    });

    it('should not remove a routine if routineId is unknow', async () => {
      const res = await Routine.remove('fake-id');
      expect(res).toBe(0);
    });
  });

  describe('validate', () => {
    it('should return true if routine is valid #1', () => {
      const routine = {
        name: 'My routine',
        sceneId: '00c1d78e-fd1c-4717-b610-65d2fa3d01b2',
        interval: '* * * * *',
      };

      expect(Routine.validate(routine)).toBeTruthy();
    });

    it('should return true if routine is valid #2', () => {
      const routine = {
        routineId: '00c1d78e-fd1c-4717-b610-65d2fa3d01b2',
        name: 'My routine',
        sceneId: '00c1d78e-fd1c-4717-b610-65d2fa3d01b2',
        createdBy: 'c10c80e8-49e4-4d6b-b966-4fc9fb98879f',
        interval: '* * * * *',
        runAtBankHoliday: true,
        createdAt: Date.now(),
      };
      expect(Routine.validate(routine)).toBeTruthy();
    });

    it('should return false if a property is unknow', () => {
      const routine = {
        routineId: '00c1d78e-fd1c-4717-b610-65d2fa3d01b2',
        name: 'My routine',
        sceneId: '00c1d78e-fd1c-4717-b610-65d2fa3d01b2',
        interval: '* * * * *',
        admin: true,
      };

      expect(Routine.validate(routine)).toBeFalsy();
    });
  });

  describe('computeNextRunAt', () => {
    it('should compute the next date the routine will execute', () => {
      isBankHoliday.mockImplementation(() => false);
      Date.now = jest
        .fn()
        .mockImplementationOnce(cb => new Date('2017-09-01T16:00').getTime());

      const routine = {
        interval: '0 12 * * *',
        runAtBankHoliday: false,
      };

      const nextRunAt = Routine.computeNextRunAt(routine);
      expect(nextRunAt).toEqual(new Date('2017-09-02T12:00'));
    });

    it('should handle bank holiday to compute the next date the routine will execute', () => {
      isBankHoliday.mockImplementation(() => true);
      Date.now = jest
        .fn()
        .mockImplementationOnce(cb => new Date('2017-08-14T16:00').getTime());

      const routine = {
        interval: '0 12 * * *',
        runAtBankHoliday: false,
      };

      const nextRunAt = Routine.computeNextRunAt(routine);
      expect(nextRunAt).toEqual(new Date('2017-08-16T12:00'));
    });
  });

  describe('run', () => {
    beforeAll(async () => {
      Routine.computeNextRunAt = jest.fn(() => new Date(2010, 5, 5));
      Routine.findByIdAndUpdate = jest.fn(() => Promise.resolve());
    });

    it('should dispatch a sceneId and call findByIdAndUpdate', async () => {
      await Routine.run(initRoutines[0]);

      expect(Routine.computeNextRunAt).toHaveBeenCalledTimes(1);
      expect(dispatch).toHaveBeenCalledWith({
        type: 'SCENE',
        id: initRoutines[0].sceneId,
      });

      expect(Routine.findByIdAndUpdate).toHaveBeenCalledWith(
        initRoutines[0].routineId,
        {
          routineId: '0fc1d78e-fd1c-4717-b610-65d2fa3d01b2',
          sceneId: '00c1d78e-fd1c-4717-b610-65d2fa3d01b2',
          name: 'test at 5am',
          interval: '0 5 * * *',
          enabled: true,
          runAtBankHoliday: true,
          createdBy: 'c10c80e8-49e4-4d6b-b966-4fc9fb98879f',
          createdAt: initRoutines[0].createdAt,
          updatedAt: expect.any(Date),
          lastRunAt: expect.any(Date),
          nextRunAt: new Date(2010, 5, 5),
          failReason: '""',
          lastFailedAt: undefined,
        },
      );
    });

    it("should return is routine don't run on bank holiday", async () => {
      isBankHoliday.mockImplementation(() => true);
      await Routine.run(initRoutines[1]);

      expect(dispatch).not.toHaveBeenCalled();
      expect(Routine.findByIdAndUpdate).not.toHaveBeenCalled();
    });
  });
});
