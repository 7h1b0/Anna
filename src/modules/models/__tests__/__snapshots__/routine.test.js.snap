// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Routines findAll should return all routines 1`] = `
Array [
  RowDataPacket {
    "createdAt": 1514764800000,
    "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
    "enabled": true,
    "failReason": null,
    "interval": "0 5 * * *",
    "lastFailedAt": 0,
    "lastRunAt": 0,
    "name": "test at 5am",
    "nextRunAt": 0,
    "routineId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
    "runAtBankHoliday": true,
    "sceneId": "00c1d78e-fd1c-4717-b610-65d2fa3d01b2",
    "updatedAt": 1514851200000,
  },
  RowDataPacket {
    "createdAt": 1514764800000,
    "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
    "enabled": true,
    "failReason": null,
    "interval": "0 9 * * *",
    "lastFailedAt": 0,
    "lastRunAt": 0,
    "name": "test at 9am",
    "nextRunAt": 0,
    "routineId": "1fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
    "runAtBankHoliday": false,
    "sceneId": "10c1d78e-fd1c-4717-b610-65d2fa3d01b2",
    "updatedAt": 1514851200000,
  },
]
`;

exports[`Routines findById should return only one routine 1`] = `
RowDataPacket {
  "createdAt": 1514764800000,
  "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
  "enabled": true,
  "failReason": null,
  "interval": "0 5 * * *",
  "lastFailedAt": 0,
  "lastRunAt": 0,
  "name": "test at 5am",
  "nextRunAt": 0,
  "routineId": "0fc1d78e-fd1c-4717-b610-65d2fa3d01b2",
  "runAtBankHoliday": true,
  "sceneId": "00c1d78e-fd1c-4717-b610-65d2fa3d01b2",
  "updatedAt": 1514851200000,
}
`;

exports[`Routines save should save a new routine 1`] = `
Object {
  "createdAt": Any<Number>,
  "createdBy": "c10c80e8-49e4-4d6b-b966-4fc9fb98879f",
  "enabled": true,
  "failReason": null,
  "interval": "schedule",
  "lastFailedAt": 0,
  "lastRunAt": 0,
  "name": "test_3",
  "nextRunAt": 0,
  "routineId": StringMatching /\\[a-fA-F0-9-\\]\\{36\\}/,
  "runAtBankHoliday": true,
  "sceneId": StringMatching /\\[a-fA-F0-9-\\]\\{36\\}/,
  "updatedAt": Any<Number>,
}
`;
